### PR TITLE
Dynamic component input value assignment

### PR DESCRIPTION
**Pull request by TheSlimvReal**
For some time we have already been planning to replace the `OnInitDynamicComponent` interface with a more Angular-like approach where we make use of the class properties marked with the `@Input()` annotation. 

This has been implemented here.

Now the components that are dynamically created just work like they are used in a template where `@Input()` properties are automatically set and available in the `ngOnInit` or even `ngOnChanges` functions.

### Visible/Frontend Changes
-- 

### Architectural/Backend Changes
- [x] `DynamicComponentDirective` automatically assigns config values to component properties annotated with `@Input()` 
- [x] removed `OnInitDynamicComponent` interface

### CODE REVIEW

File: src/app/child-dev-project/children/dashboard-widgets/entity-count-dashboard/entity-count-dashboard.component.ts [hunk 1]
```diff
@@ -31,47 +30,42 @@ import { DashboardListWidgetComponent } from "../../../../core/dashboard/dashboa
   ],
   standalone: true,
 })
-export class EntityCountDashboardComponent
-  implements OnInitDynamicComponent, OnInit
-{
-  private entity: EntityConstructor = Child;
+export class EntityCountDashboardComponent implements OnInit {
+  @Input() set entity(value: string) {
```
Comment (sleidig) [inline src/app/child-dev-project/children/dashboard-widgets/entity-count-dashboard/entity-count-dashboard.component.ts:34]: can you add a jsDoc? I am a bit confused what this does

File: src/app/child-dev-project/notes/notes-related-to-entity/notes-related-to-entity.component.ts [hunk 2]
```diff
@@ -31,12 +28,18 @@ import { EntitySubrecordComponent } from "../../../core/entity-components/entity
...
   standalone: true,
 })
-export class NotesRelatedToEntityComponent
-  implements OnChanges, OnInitDynamicComponent
-{
+export class NotesRelatedToEntityComponent implements OnInit {
   @Input() entity: Entity;
   records: Array<Note> = [];
 
+  @Input() config(config: {
```
Comment (sleidig) [inline src/app/child-dev-project/notes/notes-related-to-entity/notes-related-to-entity.component.ts:35]: mark deprecated and make `columns` and `filter` direct @ Input as well?

File: src/app/child-dev-project/schools/activities-overview/activities-overview.component.ts [hunk 3]
```diff
@@ -28,25 +27,22 @@ export class ActivitiesOverviewComponent implements OnInitDynamicComponent {
...
     },
   };
-  columns: FormFieldConfig[] = [
-    this.titleColumn,
-    { id: "type" },
-    { id: "assignedTo" },
-    { id: "linkedGroups" },
-    { id: "excludedParticipants" },
-  ];
+  @Input() config: { columns: FormFieldConfig[] } = {
```
Comment (sleidig) [inline src/app/child-dev-project/schools/activities-overview/activities-overview.component.ts:30]: flatten @ Input + legacy setter for .config?

File: src/app/features/todos/todos-related-to-entity/todos-related-to-entity.component.ts [hunk 4]
```diff
@@ -22,21 +20,22 @@ import { FormsModule } from "@angular/forms";
...
-    { id: "startDate" },
-    { id: "assignedTo" },
-    { id: "description", visibleFrom: "xl" },
-    { id: "repetitionInterval", visibleFrom: "xl" },
-    { id: "relatedEntities", hideFromTable: true },
-    { id: "completed", hideFromForm: true },
-  ];
-
   @Input() entity: Entity;
+  @Input() config: { columns: FormFieldConfig[] } = {
```
Comment (sleidig) [inline src/app/features/todos/todos-related-to-entity/todos-related-to-entity.component.ts:27]: flatten + legacy setter?

**Review by sleidig — CHANGES_REQUESTED** [review]
:heart_eyes: :heart_eyes_cat: 
I absolutely love this refactoring - feels so much clearer and can avoid duplicated and convoluted code.

I commented a few places where I would prefer to have non-nested, more clear Inputs instead of the `config` property. Maybe a two-way with the config implemented as a setter can be a workaround/fallback?

File: src/app/child-dev-project/children/children-list/recent-attendance-blocks/recent-attendance-blocks.component.ts [hunk 5]
```diff
@@ -71,32 +69,26 @@ export class RecentAttendanceBlocksComponent implements OnInitDynamicComponent {
...
 
-  async onInitFromDynamicConfig(config: ViewPropertyConfig) {
-    this.filterByActivityType = config.config.filterByActivityType;
-
-    if (config.hasOwnProperty("entity")) {
-      this.child = config.entity as Child;
-
-      let activities = await this.attendanceService.getActivitiesForChild(
-        this.child.getId()
+  async ngOnChanges() {
```
[the remark this answers is not in the document]
Comment (TheSlimvReal) [reply src/app/child-dev-project/children/children-list/recent-attendance-blocks/recent-attendance-blocks.component.ts:72]: This component has the inputs `entity` and `filterByActivityType` if either of these change, the calculation has to be repeated.

File: src/app/child-dev-project/children/children-list/bmi-block/bmi-block.component.ts [hunk 6]
```diff
@@ -23,15 +23,11 @@ export class BmiBlockComponent implements OnChanges {
...
-      .getHealthChecksOfChild(this.entity.getId())
-      .then((results) => {
-        if (results.length > 0) {
-          this.currentHealthCheck = results.reduce((prev, cur) =>
-            cur.date > prev.date ? cur : prev
-          );
-        }
-      });
+  async ngOnChanges(changes: SimpleChanges) {
+    if (changes.hasOwnProperty("entity")) {
```
Comment (TheSlimvReal) [inline src/app/child-dev-project/children/children-list/bmi-block/bmi-block.component.ts:27]: @sleidig this is not necessary, as the component only has this one input?

File: src/app/core/entity-components/entity-utils/view-components/display-date/display-date.component.ts [hunk 7]
```diff
@@ -1,26 +1,19 @@
...
 @DynamicComponent("DisplayDate")
 @Component({
   selector: "app-display-date",
-  template: `{{ value | date : format }}`,
+  template: `{{ value | date : config }}`,
   standalone: true,
   imports: [DatePipe],
 })
-export class DisplayDateComponent extends ViewDirective<Date> {
-  format = "yyyy-MM-dd";
```
[the remark this answers is not in the document]
Comment (TheSlimvReal) [reply src/app/core/entity-components/entity-utils/view-components/display-date/display-date.component.ts:18]: We used to have a default format but actually, not using a default format results in the locale format being used which I find more appropriate. What do you think?

[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/core/entity-components/entity-utils/view-components/display-date/display-date.component.ts:18]: fair point :+1:

**Comment by TheSlimvReal** [comment]
TODO:

- [ ] check if `ngOnChanges` is triggered automatically or remove where unnecessary and use `ngOnInit` instead
- [ ] flatten configs for panels in `EntityDetails` with fallback in case a `entity` name is used twice

**Pull request merged**